### PR TITLE
Switch from StrictYAML to Pydantic

### PR DESCRIPTION
**Pull request by sfc-gh-jsikorski**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Switch to Pydantic from StrictYaml. No we use Models for project definition files validation. This gives us:
* IDE-supported, safe access to definition fields
* Better control on types, names etc.
* Stricter control over input files
* No need to use hardcoded strings
* Ability to generate JSON schemas and docs from code

### CODE REVIEW

File: tests/nativeapp/test_run_processor.py [hunk 1]
```diff
@@ -794,6 +794,7 @@ def test_create_dev_app_create_new_quoted(
                 - src: app/streamlit/*.py
                 dest: ui/
 
+
```
Comment (sfc-gh-jsikorski) [inline tests/nativeapp/test_run_processor.py:797]: Is this `dest:ui/` placed here intentionally? It's an extra value, that is not used in schema? According to design doc, we should not accept extra value, so this causes an error. Should it be deleted?

File: RELEASE-NOTES.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
 ## Fixes and improvements
 * Adding `--image-name` option for image name argument in `spcs image-repository list-tags` for consistency with other commands.
 * Fixed errors during `spcs image-registry login` not being formatted correctly.
+* Validation of project definition files was switched to Pydantic, hence we no longer accept extra fields in definition files
```
Comment (sfc-gh-turbaszek) [inline RELEASE-NOTES.md:11]:
```suggestion
* Project definition no longer accept extra fields. Any extra field will cause an error.
```

File: src/snowflake/cli/api/commands/flags.py [hunk 3]
```diff
@@ -247,7 +247,7 @@ def project_definition_option(project_name: str):
 
     def _callback(project_path: Optional[str]):
         dm = DefinitionManager(project_path)
-        project_definition = dm.project_definition.get(project_name)
+        project_definition = getattr(dm.project_definition, project_name)
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/commands/flags.py:250]: We should provide default  for `getattr`

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/commands/flags.py:250]: Or implement this "get" on project definition

File: src/snowflake/cli/api/project/schemas/native_app.py [hunk 4]
```diff
@@ -1,43 +1,43 @@
...
-            {
-                Optional("role"): Regex(IDENTIFIER),
-                Optional("name"): Regex(IDENTIFIER),
-                Optional("warehouse"): Regex(IDENTIFIER),
-                Optional("debug", default=True): Bool(),
-            }
-        ),
-    }
-)
+class NativeAppSchema(UpdatableModel):
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/native_app.py:18]: I think we can consider removing `Schema` suffix. It should be fine to think about this class as representation of `NativeApp` . WDYT?

File: src/snowflake/cli/api/project/schemas/native_app_submodels.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        default=True,
+    )
+
+
+class PathMappingSchema(UpdatableModel):
+    src: str
+    dest: Optional[str] = None
+
+
+class PackageSchema(UpdatableModel):
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/native_app_submodels.py:38]: Let's keep model per file

File: src/snowflake/cli/api/project/schemas/project_definition.py [hunk 6]
```diff
@@ -1,23 +1,25 @@
...
+    native_app: Optional[NativeAppSchema] = Field(
+        title="Native app definitions for the project", default=None
+    )
+    snowpark: Optional[SnowparkSchema] = Field(
+        title="Snowpark functions and procedures definitions for the project",
+        default=None,
+    )
+    streamlit: Optional[StreamlitSchema] = Field(
+        title="Native app definitions for the project", default=None
+    )
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/project_definition.py:25]: Let's add validator for `definition_version`

File: src/snowflake/cli/api/project/schemas/snowpark.py [hunk 7]
```diff
@@ -1,47 +1,22 @@
...
-        "src": Str(),
-        Optional("functions"): Seq(function_schema),
-        Optional("procedures"): Seq(procedure_schema),
-    }
-)
+
+class SnowparkSchema(UpdatableModel):
+    project_name: str = Field(title="Project identifier")
+    stage_name: str = Field(title="Stage in which project’s artifacts will be stored")
+    src: str = Field(title="Folder where your code should be located")
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/schemas/snowpark.py:16]:
```suggestion
    src: str = Field(title="Folder where your code is located")
```

File: src/snowflake/cli/api/project/util.py [hunk 8]
```diff
@@ -4,8 +4,9 @@
 from typing import Optional
 
 IDENTIFIER = r'((?:"[^"]*(?:""[^"]*)*")|(?:[A-Za-z_][\w$]{0,254}))'
+IDENTIFIER_NO_LENGTH = r'((?:"[^"]*(?:""[^"]*)*")|(?:[A-Za-z_][\w$]*))'
 DB_SCHEMA_AND_NAME = f"{IDENTIFIER}[.]{IDENTIFIER}[.]{IDENTIFIER}"
-SCHEMA_AND_NAME = f"{IDENTIFIER}[.]{IDENTIFIER}"
+SCHEMA_AND_NAME = f"{IDENTIFIER}[.]{IDENTIFIER_NO_LENGTH}"
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/project/util.py:9]: Can you explain this change?

File: src/snowflake/cli/plugins/nativeapp/manager.py [hunk 9]
```diff
@@ -109,83 +110,106 @@ def project_root(self) -> Path:
...
         return extract_schema(self.stage_fqn)
 
     @cached_property
     def package_warehouse(self) -> Optional[str]:
-        return self.definition.get("package", {}).get("warehouse", self._conn.warehouse)
+        return (
+            self.definition.package.warehouse
+            if (self.definition.package and self.definition.package.warehouse)
+            else self._conn.warehouse
+        )
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/nativeapp/manager.py:149]: Personally I would remove all those ternary in favour of
```py
if self.definition.package and self.definition.package.warehouse:
    return self.definition.package.warehouse
return self._conn.warehouse
```
WDYT?

File: src/snowflake/cli/plugins/streamlit/commands.py [hunk 10]
```diff
@@ -86,31 +87,31 @@ def streamlit_deploy(
...
         pages_dir=Path(pages_dir),
-        stage_name=streamlit["stage"],
-        main_file=Path(streamlit["main_file"]),
+        stage_name=streamlit.stage,
+        main_file=Path(streamlit.main_file),
         replace=replace,
-        query_warehouse=streamlit["query_warehouse"],
-        additional_source_files=streamlit.get("additional_source_files"),
+        query_warehouse=streamlit.query_warehouse,
+        additional_source_files=streamlit.additional_source_files,
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/streamlit/commands.py:114]: I love this change, getting rid of all those hardcoded strings rocks 🪨

File: tests/nativeapp/test_run_processor.py (hunk 1 above)
Comment (sfc-gh-turbaszek) [reply tests/nativeapp/test_run_processor.py:797]: @sfc-gh-bgoel @sfc-gh-cgorrie can you advise?

File: tests/project/test_config.py [hunk 11]
```diff
@@ -4,29 +4,29 @@
 from unittest.mock import PropertyMock
 
 import pytest
+from pydantic import ValidationError
```
Comment (sfc-gh-turbaszek) [inline tests/project/test_config.py:7]: Should we consider raising our own?

File: tests/project/test_config.py [hunk 12]
```diff
@@ -46,36 +46,41 @@ def mock_getenv(key: str, default: Optional[str] = None) -> Optional[str]:
...
 def test_underspecified_project(project_definition_files):
-    with pytest.raises(YAMLValidationError) as exc_info:
+    with pytest.raises(ValidationError) as exc_info:
         load_project_definition(project_definition_files)
 
-    assert "required key(s) 'artifacts' not found" in str(exc_info.value)
+    assert (
+        "Field required [type=missing, input_value={'name': 'underspecified'}, input_type=dict]"
+        in str(exc_info.value)
+    )
```
Comment (sfc-gh-turbaszek) [inline tests/project/test_config.py:65]: This information should be as useful as the previous one. In current version I don't know if `artifacts` are the problem

File: tests/project/test_config.py [hunk 13]
```diff
@@ -46,36 +46,41 @@ def mock_getenv(key: str, default: Optional[str] = None) -> Optional[str]:
...
 @pytest.mark.parametrize(
     "project_definition_files", ["no_definition_version"], indirect=True
 )
 def test_fails_without_definition_version(project_definition_files):
-    with pytest.raises(YAMLValidationError) as exc_info:
+    with pytest.raises(ValidationError) as exc_info:
         load_project_definition(project_definition_files)
 
-    assert "required key(s) 'definition_version' not found" in str(exc_info.value)
+    assert "definition_version" in str(exc_info.value)
```
Comment (sfc-gh-turbaszek) [inline tests/project/test_config.py:75]: Let's make longer assertion so we can see the error

File: tests/project/test_config.py [hunk 14]
```diff
@@ -46,36 +46,41 @@ def mock_getenv(key: str, default: Optional[str] = None) -> Optional[str]:
...
-    project = load_project_definition(project_definition_files)
-    assert project["native_app"]["name"] == "unknown_fields"
-    assert project["native_app"]["unknown_fields_accepted"] == True
+def test_does_not_accept_unknown_fields(project_definition_files):
+    with pytest.raises(ValidationError) as e:
+        project = load_project_definition(project_definition_files)
+
+    assert (
+        "Extra inputs are not permitted [type=extra_forbidden, input_value=True, input_type=bool]"
+        in e.value.__str__()
```
Comment (sfc-gh-turbaszek) [inline tests/project/test_config.py:85]: We should make sure this more meaningful message

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 6 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/project_definition.py:25]: you mean `definition == 1`?

File: src/snowflake/cli/api/project/schemas/native_app.py (hunk 4 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app.py:18]: Done

File: src/snowflake/cli/api/project/schemas/native_app_submodels.py (hunk 5 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app_submodels.py:38]: Done

File: src/snowflake/cli/plugins/nativeapp/manager.py (hunk 9 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/plugins/nativeapp/manager.py:149]: Done

File: src/snowflake/cli/api/project/schemas/snowpark.py (hunk 7 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/snowpark.py]: Done

File: src/snowflake/cli/api/project/util.py (hunk 8 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/util.py]: That was a lost artifact of long battle with regexes. Long story.

File: tests/project/test_config.py (hunk 11 above)
Comment (sfc-gh-jsikorski) [reply tests/project/test_config.py]: Implemented

File: tests/project/test_config.py (hunk 13 above)
Comment (sfc-gh-jsikorski) [reply tests/project/test_config.py]: Done

**Comment by sfc-gh-cgorrie** [comment]
@sfc-gh-jsikorski just thinking a bit ahead to IDE integration. It looks like pydantic has the ability to generate JSON schema, which we'd like to consume inside of IDEs. Eventually we should extract the project definition schemas into their own open-source project which can also host these JSON schemas as part of a CD process.

Wanted to confirm this medium-term vision is compatible with your future plans too!

File: src/snowflake/cli/api/project/definition_manager.py [hunk 15]
```diff
@@ -100,5 +101,5 @@ def _user_definition_file_if_available(project_path: Path) -> Optional[Path]:
         )
 
     @functools.cached_property
-    def project_definition(self) -> dict:
+    def project_definition(self) -> ProjectDefinition:
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/definition_manager.py:104]: Love the type safety Pydantic gives us!

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 6 above)
Comment (sfc-gh-cgorrie) [reply src/snowflake/cli/api/project/schemas/project_definition.py:25]: Another way to think about this -- when we get to project definition v2 (design currently in progress) it will look completely different. We should have a way to separate out the definitions by `definition_version` rather than have to have a complex base model that has to support both. I agree that for now ensuring `definition_version == 1` is important, but we should make sure that when we introduce `definition_version == 2` it's something that can live alongside this model rather than "within" it.

File: tests/project/test_config.py (hunk 12 above)
Comment (sfc-gh-jsikorski) [reply tests/project/test_config.py]: Implemented

File: src/snowflake/cli/api/project/schemas/native_app/path_maping.py [hunk 16]
```diff
@@ -0,0 +1,10 @@
+from __future__ import annotations
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/native_app/path_maping.py:1]: This filename has a typo (should be `path_mapping.py`).

File: tests/nativeapp/test_run_processor.py (hunk 1 above)
Comment (sfc-gh-cgorrie) [reply tests/nativeapp/test_run_processor.py:796]: I think this is an indentation error -- this is part of the last entry of the artifacts collection (i.e. `{ src: "app/streamlit/*.py", dest: "ui/" }`).

**Review by sfc-gh-cgorrie — COMMENTED** [review]
Some initial questions and comments while I dive deeper into the code.

File: src/snowflake/cli/api/project/definition_manager.py (hunk 15 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/definition_manager.py:104]: Me too :)
As i cannot answer the comment above - yes, Pydantic can be used to generate JSON schemas, but i think i'll summon @sfc-gh-turbaszek to answer the long-term vision plans :D

Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/definition_manager.py:104]: And we could use `description` and `examples` fields of FieldInfo, to have the schema well documented.

File: src/snowflake/cli/api/project/schemas/native_app/path_maping.py (hunk 16 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app/path_maping.py:1]: Done

File: RELEASE-NOTES.md [hunk 17]
```diff
@@ -10,6 +10,8 @@
 ## Fixes and improvements
 * Adding `--image-name` option for image name argument in `spcs image-repository list-tags` for consistency with other commands.
 * Fixed errors during `spcs image-registry login` not being formatted correctly.
+* Project definition no longer accept extra fields. Any extra field will cause an error.
+* Project definition no longer accept extra fields. Any extra field will cause an error.
```
Comment (sfc-gh-cgorrie) [inline RELEASE-NOTES.md:14]: Duplicate line

File: src/snowflake/cli/api/project/schemas/native_app/package.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+class Package(UpdatableModel):
+    scripts: Optional[List[str]] = Field(
+        title="List of SQL file paths relative to the project root", default=None
+    )
+    role: Optional[str] = IdentifierField(
+        title="Role to use when creating the application package and provider-side objects",
+        default=None,
+    )
+    name: Optional[str] = IdentifierField(
+        title="Name of the application created when you run the snow app run command",  # TODO: this description seems duplicated, is it ok?
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/native_app/package.py:22]:
```suggestion
        title="Name of the application package created when you run the snow app run command",
```

File: src/snowflake/cli/api/project/schemas/native_app/package.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+        title="Distribution of the application package created by the Snowflake CLI",
+        default="internal",
+    )
+
+    @field_validator("scripts")
+    @classmethod
+    def validate_scripts(cls, input_list):
+        if len(input_list) != len(set(input_list)):
+            raise ValueError(
+                "Scripts field should contain unique values. Check the list for duplicates and try again"
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/native_app/package.py:38]:
```suggestion
                "package.scripts field should contain unique values. Check the list for duplicates and try again"
```

Comment (sfc-gh-cgorrie) [reply src/snowflake/cli/api/project/schemas/native_app/package.py:38]: Suggesting this change because we may add "scripts" keys to other parts of the schema

**Review by sfc-gh-cgorrie — COMMENTED** [review]
Some more comments and suggestions.

Also I have a question, is it intentional that we have a `test_pydantic_schemas.py` that is empty?

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 6 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/project_definition.py:25]: Ok, so my proposition is to add `definition_version==1` to field validation. If user provides another version of definition, it's pointless to proceed with parsing any fields.
Then, when we add new versions, we can change `load_project_definition()` logic to decide which model should be used for provided data.

Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/schemas/project_definition.py:25]: I agree, let's have `==1` for now. We will add more validation once we start working on v2.

**Comment by sfc-gh-turbaszek** [comment]
> @sfc-gh-jsikorski just thinking a bit ahead to IDE integration. It looks like pydantic has the ability to generate JSON schema, which we'd like to consume inside of IDEs. Eventually we should extract the project definition schemas into their own open-source project which can also host these JSON schemas as part of a CD process.
> 
> Wanted to confirm this medium-term vision is compatible with your future plans too!

The ability to generate JSON schema is one of the main goals (IDE, docs, etc). In terms of vision - extracting this to separate package is not planned yet, but once we have py ecosystem then we can easily do so.

File: pyproject.toml [hunk 20]
```diff
@@ -25,6 +25,7 @@ dependencies = [
   "typer==0.9.0",
   "urllib3>=1.21.1,<2.3",
   "GitPython==3.1.42",
+  "pydantic==2.6.3"
```
Comment (sfc-gh-turbaszek) [inline pyproject.toml:28]: Should we remove need for strictyml? Or should it be done in separate PR?

File: src/snowflake/cli/api/project/schemas/native_app/package.py (hunk 19 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app/package.py:38]: Done

File: RELEASE-NOTES.md (hunk 17 above)
Comment (sfc-gh-jsikorski) [reply RELEASE-NOTES.md:14]: Done

File: src/snowflake/cli/api/project/schemas/project_definition.py (hunk 6 above)
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/project_definition.py:27]: Added

**Review by sfc-gh-turbaszek — DISMISSED** [review]

**Review by sfc-gh-bgoel — COMMENTED** [review]
The main changes LGTM. Left some comments about messages shown to the user which need to be in accordance with our PMM team's guidelines.

File: src/snowflake/cli/api/project/schemas/native_app/native_app.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+        default="app_src.stage",
+    )
+    package: Optional[Package] = Field(title="PackageSchema", default=None)
+    application: Optional[Application] = Field(title="Application info", default=None)
+
+    @field_validator("source_stage")
+    @classmethod
+    def validate_source_stage(cls, input_value: str):
+        if not re.match(SCHEMA_AND_NAME, input_value):
+            raise ValueError("Incorrect value for Native Apps source stage value")
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/project/schemas/native_app/native_app.py:38]: +1, I think latter one we can add something like `regex_validator(value, pattern)` that would cause a generic error like `Incorrect value for native_app.source_stage. Allowed values: pattern`.

File: src/snowflake/cli/api/project/schemas/native_app/application.py [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+from pydantic import Field
+from snowflake.cli.api.project.schemas.updatable_model import (
+    IdentifierField,
+    UpdatableModel,
+)
+
+
+class Application(UpdatableModel):
+    role: Optional[str] = Field(
+        title="Role to use when creating the application instance and consumer-side objects",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app/application.py:14]: Done

File: src/snowflake/cli/api/project/schemas/native_app/application.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+        default=None,
+    )
+    name: Optional[str] = Field(
+        title="Name of the application created when you run the snow app run command",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app/application.py:18]: Done

File: src/snowflake/cli/api/project/schemas/native_app/application.py [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+        default=None,
+    )
+    warehouse: Optional[str] = IdentifierField(
+        title="Name of the application created when you run the snow app run command",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app/application.py:22]: Done

File: src/snowflake/cli/api/project/schemas/native_app/native_app.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jsikorski) [reply src/snowflake/cli/api/project/schemas/native_app/native_app.py]: Done

**Review by sfc-gh-bgoel — APPROVED** [review]
LGTM for native app and project changes, thank you!

**Review by sfc-gh-turbaszek — APPROVED** [review]
